Moved publishing of code coverage artifacts to test_script phase

adityapatwardhan opened the pull request:
The webhook for build completion is called after the test_script phase
hence we needed to publish code coverage artifacts before that. Moved the
logic for compression and publish the artifacts after running tests.

adityapatwardhan committed:
Moved publishing of code coverage artifacts to after_test phase

The webhook for build completion is called after the after_test phase
hence we needed to publish code coverage artifacts before that. Moved the
logic for compression and publish the artifacts after running tests.

adityapatwardhan commented:
@daxian-dbw Can you have a look at this? The call needed to be moved as per appveyor documentation: https://www.appveyor.com/docs/build-configuration/ for build pipeline.

adityapatwardhan committed:
Added check for daily build

vors commented on tools/appveyor.psm1:
```diff
@@ -342,6 +348,32 @@ function Invoke-AppVeyorTest
+function Compress-CoverageArtifacts
+{
+    # Create archive for test content, OpenCover module and CodeCoverage build
+    if(Test-DailyBuild)
```
This if seems unneeded. Whoever call `Compress-CoverageArtifacts` should already know what they are doing.

adityapatwardhan replied:
Removed the check.

---
In reply to: [89253173](https://github.com/PowerShell/PowerShell/pull/2760#discussion_r89253173) [](ancestors = 89253173)

vors commented on tools/appveyor.psm1:
```diff
@@ -326,9 +326,15 @@ function Invoke-AppVeyorTest
- 
+    if(Test-DailyBuild)
+    {
+        ## Publish code coverage build, tests and OpenCover module to artifacts, so webhook has the information.
```
> artifacts

I think the comment is missleading. AppVeyor has separate tabs/concepts for Tests and Artifacts. It seems that here we use `Update-AppVeyorTestResults` => tests tab and don't publish the artifacts (in the appveyor terminology).

adityapatwardhan replied:
Resolved. Using Push-AppVeyorArtifact instead.

---
In reply to: [89253329](https://github.com/PowerShell/PowerShell/pull/2760#discussion_r89253329) [](ancestors = 89253329)

vors commented on tools/appveyor.psm1:
```diff
@@ -326,9 +326,15 @@ function Invoke-AppVeyorTest
+    {
+        ## Publish code coverage build, tests and OpenCover module to artifacts, so webhook has the information.
+        ## Build webhook is called after 'after_test' phase, hence we need to do this here and not in AppveyorFinish. 
+        $env:CodeCoverageOutput = Split-Path -Parent (Get-PSOutput -Options (New-PSOptions -Configuration CodeCoverage))
```
Do we really need `env` variable? It's not the best way to pass parameters around and should be avoided, if possible.

adityapatwardhan replied:
Was an left over artifact from the previous design. Changed.

---
In reply to: [89253374](https://github.com/PowerShell/PowerShell/pull/2760#discussion_r89253374) [](ancestors = 89253374)

adityapatwardhan committed:
Address code review comments.

daxian-dbw commented on tools/appveyor.psm1:
```diff
@@ -326,9 +326,15 @@ function Invoke-AppVeyorTest
-    ## CodeCoverage
-    $env:CodeCoverageOutput = Split-Path -Parent (Get-PSOutput -Options (New-PSOptions -Configuration CodeCoverage))
- 
+    if(Test-DailyBuild)
```
Chat with @adityapatwardhan offline and we think this logic can be wrapped into a function and be called in the `after_test:` phase.

daxian-dbw commented on tools/appveyor.psm1:
```diff
@@ -326,9 +326,15 @@ function Invoke-AppVeyorTest
+    {
+        ## Publish code coverage build, tests and OpenCover module to artifacts, so webhook has the information.
+        ## Build webhook is called after 'after_test' phase, hence we need to do this here and not in AppveyorFinish. 
+        $CodeCoverageOutput = Split-Path -Parent (Get-PSOutput -Options (New-PSOptions -Configuration CodeCoverage))
```
Minor: maybe change `$CodeCoverageOutput` to `$codeCoverageOutput` with lower `c`, so that it's in the same format as the variables in the context.

adityapatwardhan replied:
Changed.

daxian-dbw commented on tools/appveyor.psm1:
```diff
@@ -342,6 +348,32 @@ function Invoke-AppVeyorTest
+
+    $zipTestContentPath = Join-Path $pwd 'tests.zip'
+    Compress-TestContent -Destination $zipTestContentPath
+    $artifacts.Add($zipTestContentPath)
```
This will cause an exception because you are adding to an array:
`Exception calling "Add" with "1" argument(s): "Collection was of a fixed size."`

adityapatwardhan replied:
Defined $artifacts as ArrayList

adityapatwardhan committed:
Addressed code review comments

Addressed comments about definition of artifacts.
Defined new function for after_test phase.
